mlbench_core/dataset/util/pytorch/libsvm.py

```diff
@@ -0,0 +1,226 @@
+        'sparse': True},
+    'rcv1_test': {
+        'url': 'https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets'
+               '/binary/rcv1_test.binary.bz2',
+        'sparse': True},
+    'webspam_train': {
+        'url': 'https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary'
+               '/webspam_wc_normalized_trigram.svm.bz2',
+        'sparse': True
+    }
+
+}
+
+
+def _get_dense_tensor(tensor):
+    if 'sparse' in str(type(tensor)):
+        return tensor.toarray()
+    elif 'numpy' in str(type(tensor)):
+        return tensor
+
+
+def _correct_binary_labels(labels, is_01_classes=True):
+    classes = set(labels)
+
+    if -1 in classes and is_01_classes:
+        labels[labels == -1] = 0
+    return labels
+
+
+class LIBSVMDataset(object):
+    def __init__(self, root_path, name, data_type):
+
+        # get file url and file path.
+        data = _DATASET_MAP['{}_{}'.format(name, data_type)]
+        data_url = data['url']
+        self.is_sparse = data['sparse']
+
+        file_name = data_url.split('/')[-1]
+        # Downloads and extracts the data (deletes downloaded file after)
+        file_path = maybe_download_and_extract_bz2(root_path, file_name,
+                                                   data_url)
+
+        print("Loading SVM file {}".format(file_path))
+        data = load_svmlight_file(file_path)
+        self.features, self.labels = self._get_features_and_labels(data)
+
+    def _get_features_and_labels(self, data):
+        features, labels = data
+
+        features = _get_dense_tensor(features) if not self.is_sparse else \
+            features
+        labels = _get_dense_tensor(labels)
+        labels = _correct_binary_labels(labels)
+        return features, labels
+
+    def __len__(self):
+        return self.features.shape[0]
+
+    def __iter__(self):
+        idxs = list(range(self.__len__()))
+        for k in idxs:
+            features = self.features[k]
+            label = self.labels[k]
+            if self.is_sparse:
+                features = features.tocoo()
+                yield [(features.data, features.row, features.col), label]
+            else:
+                yield [features, label]
+
+    def get_data(self):
+        return self.__iter__()
+
+    def size(self):
+        return self.__len__()
+
+    def reset_state(self):
+        """
+        Reset state of the dataflow.
+        It **has to** be called once and only once before producing datapoints.
+        Note:
+            1. If the dataflow is forked, each process will call this method
+               before producing datapoints.
+            2. The caller thread of this method must remain alive to keep
+            this dataflow alive.
+        For example, RNG **has to** be reset if used in the DataFlow,
+        otherwise it won't work well with prefetching, because different
+        processes will have the same RNG state.
+        """
+        pass
+
+
+class SyntheticLIBSVMDataset(object):
```

Comment:
This could inherit from `LIBSVMDataset` so we don't need to duplicate `len`, `get_data`, `size` and `reset_state`

Reply:
But the constructor is different, for both classes. I can create an parent class to both so they inherit common functions